flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBuffer.java

```diff
@@ -0,0 +1,494 @@
+		this.mode = mode;
+		this.operator = operator;
+
+		this.lock = new ReentrantLock(true);
+		this.notFull = this.lock.newCondition();
+		this.taskDone = this.lock.newCondition();
+		this.isEmpty = this.lock.newCondition();
+
+		this.emitter = new Emitter();
+		this.emitThread = new Thread(emitter);
+	}
+
+	/**
+	 * Add an {@link StreamRecord} into the buffer. A new {@link AsyncCollector} will be created and returned
+	 * corresponding to the input StreamRecord.
+	 * <p>
+	 * If buffer is full, caller will wait until new space is available.
+	 *
+	 * @param record StreamRecord
+	 * @return An AsyncCollector
+	 * @throws Exception InterruptedException or exceptions from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> add(StreamRecord<IN> record) throws Exception {
+		try {
+			lock.lock();
+
+			notifyCheckpointDone();
+
+			while (queue.size() >= bufferSize) {
+				notFull.await();
+			}
+
+			// propagate error to the main thread
+			if (error != null) {
+				throw error;
+			}
+
+			AsyncCollector<IN, OUT> collector = new AsyncCollector(this);
+
+			collectorToQueue.put(collector, queue.add(collector));
+			collectorToStreamElement.put(collector, record);
+
+			return collector;
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	/**
+	 * Add a {@link Watermark} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will be blocked here.
+	 *
+	 * @param watermark Watermark
+	 * @return AsyncCollector
+	 * @throws Exception Exceptions from async operation.
+	 */
+	public AsyncCollector<IN, OUT> add(Watermark watermark) throws Exception {
+		return processMark(watermark);
+	}
+
+	/**
+	 * Add a {@link LatencyMarker} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will be blocked here.
+	 *
+	 * @param latencyMarker LatencyMarker
+	 * @return AsyncCollector
+	 * @throws Exception Exceptions from async operation.
+	 */
+	public AsyncCollector<IN, OUT> add(LatencyMarker latencyMarker) throws Exception {
+		return processMark(latencyMarker);
+	}
+
+	private AsyncCollector<IN, OUT> processMark(StreamElement mark) throws Exception {
+		try {
+			lock.lock();
+
+			notifyCheckpointDone();
+
+			while (queue.size() >= bufferSize)
+				notFull.await();
+
+			if (error != null) {
+				throw error;
+			}
+
+			AsyncCollector<IN, OUT> collector = new AsyncCollector(this, true);
+
+			collectorToQueue.put(collector, queue.add(collector));
+			collectorToStreamElement.put(collector, mark);
+
+			// signal emitter thread that current collector is ready
+			mark(collector);
+
+			return collector;
+		}
+		finally {
+			lock.unlock();
+		}
+	}
+
+	/**
+	 * Notify the Emitter Thread that an AsyncCollector has completed.
+	 *
+	 * @param collector Completed AsyncCollector
+	 */
+	void mark(AsyncCollector<IN, OUT> collector) {
+		try {
+			lock.lock();
+
+			if (mode == AsyncDataStream.OutputMode.UNORDERED) {
+				finishedCollectors.add(collector);
```

Comment:
Yep, I will try to throw an exception while compiling the job graph for this case.